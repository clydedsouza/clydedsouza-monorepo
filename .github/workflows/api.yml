name: Build and deploy API

on:
  push:
    branches:
      - main
    paths:
      - "api/**"
  pull_request:
    branches:
      - main
    paths:
      - "api/**"
  workflow_dispatch:

jobs:
  build_api:
    name: Build API
    runs-on: ubuntu-latest
    environment:
      name: api
    defaults:
      run:
        shell: bash
        working-directory: api

    steps:
      - name: Checkout code
        uses: actions/checkout@v3

      - name: Install Node.js
        uses: actions/setup-node@v3
        with:
          node-version: 18.x

      - name: Install NPM packages
        run: npm ci

      - name: Generate API files
        run: npm run generate:api

      - name: Upload API files
        uses: actions/upload-artifact@v4
        with:
          path: ./api/public
          include-hidden-files: true

  deploy_api:
    name: Deploy API
    runs-on: ubuntu-latest
    environment:
      name: api
    needs: build_api
    if: |
      success() &&
      github.ref == 'refs/heads/main'
    defaults:
      run:
        shell: bash

    steps:
      - name: Download artifact
        uses: actions/download-artifact@v4
        with:
          name: artifact
          path: artifact

      - name: Deploy artifacts to the API repo
        uses: tagus/git-deploy@v0.4.1
        with:
          changes: artifact
          repository: git@github.com:clydedsouza/clydedsouza-web-api.git
          ssh_key: ${{ secrets.API_REPO_DEPLOY_KEY }}
          name: clydedsouza
          email: 16275549+clydedsouza@users.noreply.github.com
          clean_repo: true
          branch: main
          message: "Commit from the monorepo API deploy job"
