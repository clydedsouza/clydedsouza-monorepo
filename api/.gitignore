# Dependency directories
node_modules/

# Optional npm cache directory
.npm

# Exclude temp 'working' directory
public/*.json