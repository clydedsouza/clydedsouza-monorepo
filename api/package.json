{
  "name": "clydedsouza-projects",
  "version": "0.0.1",
  "author": "Clyde D'Souza",
  "scripts": {
    "generate:portfolio": "markdown-json -c ./settings/portfolio.json",
    "generate:projects": "markdown-json -c ./settings/projects.json",
    "generate:books": "markdown-json -c ./settings/books.json",    
    "generate:speaking": "markdown-json -c ./settings/speaking.json",
    "generate:teaching": "markdown-json -c ./settings/teaching.json",
    "generate:highlights": "markdown-json -c ./settings/highlights.json",
    "generate:pinned": "markdown-json -c ./settings/pinned.json",
    "generate:api": "npm run generate:projects & npm run generate:teaching & npm run generate:books & npm run generate:speaking & npm run generate:highlights & npm run generate:pinned & npm run generate:portfolio"
  },
  "dependencies": {
    "markdown-json": "^1.6.1"
  }
}
