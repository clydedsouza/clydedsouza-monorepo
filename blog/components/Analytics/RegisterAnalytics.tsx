'use client'

import { useEffect } from 'react'
import { initMixpanel } from '../../lib/mixpanel'

export const RegisterAnalytics = () => {
  useEffect(() => {
    initMixpanel()
  }, [])
  return <></>
}
