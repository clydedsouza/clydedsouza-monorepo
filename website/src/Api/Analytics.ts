import mixpanel from 'mixpanel-browser'
import { PageTypes } from '../Components/Navigation/PageTypes'
import {
  ILinkClickedProps,
  MIXPANEL_DEV_PROJECT_ID,
  MIXPANEL_PROD_PROJECT_ID,
} from './IAnalytics'

export const initAnalyticsWithSuperProperties = () => {
  const mixpanelProjectId =
    process.env.NODE_ENV === 'production'
      ? MIXPANEL_PROD_PROJECT_ID
      : MIXPANEL_DEV_PROJECT_ID

  try {
    mixpanel.init(mixpanelProjectId)
    mixpanel.register({
      environment: process.env.NODE_ENV,
    })
  } catch {
    undefined
  }
}

export const sendPageViewEvent = (pageType: PageTypes) => {
  try {
    mixpanel.track('Page view', { page: pageType })
  } catch {
    undefined
  }
}

export const sendLinkClickedEvent = (props: ILinkClickedProps) => {
  try {
    mixpanel.track('Link clicked', {
      destination_url: props.link,
      link_type: props.type,
      page_location: props.location,
    })
  } catch {
    undefined
  }
}
