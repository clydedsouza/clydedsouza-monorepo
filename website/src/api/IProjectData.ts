export interface IProjectData {
  data: IProject[]
}

export interface IProject {
  id: string
  title: string
  date: string
  description: string
  website: string
  github: string
  madeUsing: string[]
  category: string
  image: string
  imageDescription: string
}
