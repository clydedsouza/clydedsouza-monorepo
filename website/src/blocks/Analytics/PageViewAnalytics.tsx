"use client";

import { sendPageViewEvent } from "@/api/Analytics";
import { useEffect } from "react";
import { PageTypes } from "../Navigation/PageTypes";

export const PageViewAnalytics: React.FC<{ page: PageTypes }> = ({ page }) => {
  useEffect(() => {
    sendPageViewEvent(page);
  }, []);
  return <></>;
};
