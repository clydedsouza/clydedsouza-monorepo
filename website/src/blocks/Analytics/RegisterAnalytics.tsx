"use client";

import { initAnalyticsWithSuperProperties } from "@/api/Analytics";
import { useEffect } from "react";

export const RegisterAnalytics = () => {
  useEffect(() => {
    initAnalyticsWithSuperProperties();
  }, []);
  return <></>;
};
