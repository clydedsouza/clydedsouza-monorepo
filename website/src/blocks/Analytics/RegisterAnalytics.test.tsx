import * as analyticsModule from "@/api/Analytics";
import { render } from "@testing-library/react";
import { RegisterAnalytics } from "./RegisterAnalytics";

jest.mock("../../api/Analytics");

describe("Register Analytics", () => {
  it("should render component", () => {
    const { container } = render(<RegisterAnalytics />);
    expect(container).toMatchSnapshot();
  });

  it("should initialize analytics when component is rendered", async () => {
    const initAnalyticsWithSuperPropertiesMock = jest
      .spyOn(analyticsModule, "initAnalyticsWithSuperProperties")
      .mockImplementation(jest.fn());

    render(<RegisterAnalytics />);

    expect(initAnalyticsWithSuperPropertiesMock).toHaveBeenCalledTimes(1);
  });
});
