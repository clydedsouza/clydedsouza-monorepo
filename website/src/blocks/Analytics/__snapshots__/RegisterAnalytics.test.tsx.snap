// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Analytics should render component 1`] = `<div />`;
